File: backend/api/rbac.py
"""
Hotel ERP RBAC — roles mapped to least-privilege modules and capabilities.

Designed around global PMS practice (Opera / Mews / Cloudbeds patterns):
- Roles = stable job functions, not one-off titles
- Tenant modules ∩ role modules = effective access
- Separation of duties: Purchase Officer cannot post GL vouchers;
  Accountant owns ledgers; Front Desk owns guest folio ops
"""
from __future__ import annotations

from typing import Any

# Capability flags used by UI + future API guards
CAP_VIEW_DASHBOARD = 'dashboard.view'
CAP_MANAGE_USERS = 'users.manage'
CAP_POST_VOUCHERS = 'accounts.post_vouchers'
CAP_MANAGE_COA = 'accounts.manage_coa'
CAP_VIEW_FINANCE = 'accounts.view_reports'
CAP_PURCHASE = 'inventory.purchase'
CAP_APPROVE_PO = 'inventory.approve_po'
CAP_POS = 'fnb.pos'
CAP_CHECKIN = 'frontdesk.checkin'
CAP_FOLIO = 'frontdesk.folio'
CAP_ROOM_STATUS = 'housekeeping.room_status'
CAP_HK_TASKS = 'housekeeping.tasks'

ALL_MODULE_KEYS = [
    'frontdesk',
    'housekeeping',
    'fnb',
    'recipes',
    'laundry',
    'spa',
    'hall',
    'banquet',
    'pool',
    'crm',
    'accounts',
    'inventory',
    'assets',
    'broadcast',
    'hr',
    'channel',
    'reports',
    'utilities',
    'landing',
]

# Menu section titles (must match gyoroom-menu top-level titles)
ROLE_DEFINITIONS: dict[str, dict[str, Any]] = {
    'superadmin': {
        'label': 'Platform Superadmin',
        'description': 'SaaS control plane — tenants, billing, catalog.',
        'home_path': '/saas',
        'modules': list(ALL_MODULE_KEYS),
        'menu_sections': None,  # all
        'capabilities': [CAP_MANAGE_USERS, CAP_VIEW_DASHBOARD],
        'color': '#6366f1',
    },
    'admin': {
        'label': 'Property Administrator',
        'description': 'Full property access including utilities and user management.',
        'home_path': '/home',
        'modules': list(ALL_MODULE_KEYS),
        'menu_sections': None,
        'capabilities': [
            CAP_VIEW_DASHBOARD,
            CAP_MANAGE_USERS,
            CAP_POST_VOUCHERS,
            CAP_MANAGE_COA,
            CAP_VIEW_FINANCE,
            CAP_PURCHASE,
            CAP_APPROVE_PO,
            CAP_POS,
            CAP_CHECKIN,
            CAP_FOLIO,
            CAP_ROOM_STATUS,
            CAP_HK_TASKS,
        ],
        'color': '#0f766e',
    },
    'operations_manager': {
        'label': 'Operations Manager',
        'description': 'Cross-department ops: rooms, F&B, inventory, assets, CRM, reports. No user admin.',
        'home_path': '/home/operations',
        'modules': [
            'frontdesk',
            'housekeeping',
            'fnb',
            'recipes',
            'laundry',
            'spa',
            'hall',
            'banquet',
            'pool',
            'crm',
            'inventory',
            'assets',
            'broadcast',
            'reports',
            'channel',
        ],
        'menu_sections': [
            'Dashboard',
            'FRONTDESK',
            'HOUSEKEEPING',
            'BANQUET',
            'F&B AND REVENUE CENTER',
            'INVENTORY',
            'SALES & MARKETING',
            'ASSET & MAINTANANCE',
            'ASSET & MAINTENANCE',
            'BROADCAST MESSAGE',
            'CHANNEL MANAGER',
            'REPORT CENTER',
        ],
        'capabilities': [
            CAP_VIEW_DASHBOARD,
            CAP_PURCHASE,
            CAP_APPROVE_PO,
            CAP_POS,
            CAP_CHECKIN,
            CAP_FOLIO,
            CAP_ROOM_STATUS,
            CAP_HK_TASKS,
            CAP_VIEW_FINANCE,
        ],
        'color': '#1d4ed8',
    },
    'manager': {
        # Alias kept for legacy users — same as operations_manager
        'label': 'Manager',
        'description': 'Property operations oversight (legacy alias of Operations Manager).',
        'home_path': '/home/operations',
        'modules': None,  # copy from operations_manager at resolve time
        'alias_of': 'operations_manager',
        'color': '#1d4ed8',
    },
    'frontdesk': {
        'label': 'Front Desk',
        'description': 'Reservations, arrivals/departures, folio, room rack, guest CRM.',
        'home_path': '/home/frontdesk',
        'modules': ['frontdesk', 'housekeeping', 'crm', 'broadcast', 'reports', 'channel'],
        'menu_sections': [
            'Dashboard',
            'FRONTDESK',
            'HOUSEKEEPING',
            'SALES & MARKETING',
            'BROADCAST MESSAGE',
            'CHANNEL MANAGER',
            'REPORT CENTER',
        ],
        'capabilities': [
            CAP_VIEW_DASHBOARD,
            CAP_CHECKIN,
            CAP_FOLIO,
            CAP_ROOM_STATUS,
        ],
        'color': '#2563eb',
    },
    'housekeeping': {
        'label': 'Housekeeping',
        'description': 'Room status board, HK tasks, lost & found, amenity, laundry.',
        'home_path': '/home/housekeeping',
        'modules': ['housekeeping', 'laundry', 'assets', 'reports'],
        'menu_sections': [
            'Dashboard',
            'HOUSEKEEPING',
            'LAUNDRY',
            'ASSET & MAINTANANCE',
            'ASSET & MAINTENANCE',
            'REPORT CENTER',
        ],
        'capabilities': [CAP_VIEW_DASHBOARD, CAP_ROOM_STATUS, CAP_HK_TASKS],
        'color': '#059669',
    },
    'restaurant': {
        'label': 'Restaurant / F&B',
        'description': 'POS, orders, menu, revenue centers, recipes, kitchen stock.',
        'home_path': '/home/restaurant',
        'modules': ['fnb', 'recipes', 'inventory', 'reports'],
        'menu_sections': [
            'Dashboard',
            'F&B AND REVENUE CENTER',
            'FOOD & BEVERAGE',
            'RECIPE MANAGEMENT',
            'INVENTORY',
            'REPORT CENTER',
        ],
        'capabilities': [CAP_VIEW_DASHBOARD, CAP_POS],
        'color': '#c2410c',
    },
    'fnb': {
        'label': 'F&B',
        'description': 'Legacy F&B role — same access as Restaurant.',
        'home_path': '/home/restaurant',
        'alias_of': 'restaurant',
        'color': '#c2410c',
    },
    'accountant': {
        'label': 'Accountant',
        'description': 'Chart of accounts, vouchers, ledgers, financial reports. No purchasing.',
        'home_path': '/home/accountant',
        'modules': ['accounts', 'reports'],
        'menu_sections': ['Dashboard', 'ACCOUNTS', 'REPORT CENTER'],
        'capabilities': [
            CAP_VIEW_DASHBOARD,
            CAP_POST_VOUCHERS,
            CAP_MANAGE_COA,
            CAP_VIEW_FINANCE,
        ],
        'color': '#7c3aed',
        'account_permissions': {
            'can_post_vouchers': True,
            'can_view_reports': True,
            'can_manage_coa': True,
        },
    },
    'purchase_officer': {
        'label': 'Purchase Officer',
        'description': 'Purchase orders, suppliers, stock receipts. Cannot post GL vouchers (SoD).',
        'home_path': '/home/purchase',
        'modules': ['inventory', 'reports'],
        'menu_sections': ['Dashboard', 'INVENTORY', 'REPORT CENTER'],
        'capabilities': [CAP_VIEW_DASHBOARD, CAP_PURCHASE],
        'color': '#b45309',
        'account_permissions': {
            'can_post_vouchers': False,
            'can_view_reports': True,
            'can_manage_coa': False,
        },
    },
    'maintenance': {
        'label': 'Maintenance',
        'description': 'Assets, work orders, room out-of-order coordination.',
        'home_path': '/home/housekeeping',
        'modules': ['assets', 'housekeeping', 'reports'],
        'menu_sections': [
            'Dashboard',
            'HOUSEKEEPING',
            'ASSET & MAINTANANCE',
            'ASSET & MAINTENANCE',
            'REPORT CENTER',
        ],
        'capabilities': [CAP_VIEW_DASHBOARD, CAP_ROOM_STATUS],
        'color': '#64748b',
    },
    'staff': {
        'label': 'Staff',
        'description': 'Minimal dashboard access until assigned a department role.',
        'home_path': '/home',
        'modules': [],
        'menu_sections': ['Dashboard'],
        'capabilities': [CAP_VIEW_DASHBOARD],
        'color': '#6b7280',
    },
}


def resolve_role(role: str | None) -> dict[str, Any]:
    key = (role or 'staff').lower().strip()
    base = ROLE_DEFINITIONS.get(key) or ROLE_DEFINITIONS['staff']
    if base.get('alias_of'):
        parent = ROLE_DEFINITIONS[base['alias_of']]
        merged = {**parent, **{k: v for k, v in base.items() if k not in ('alias_of', 'modules', 'menu_sections', 'capabilities')}}
        merged['modules'] = parent.get('modules')
        merged['menu_sections'] = parent.get('menu_sections')
        merged['capabilities'] = parent.get('capabilities')
        merged['key'] = key
        return merged
    out = dict(base)
    out['key'] = key
    return out


def effective_modules(role: str | None, tenant_modules: list[str] | None, *, is_superuser: bool = False) -> list[str]:
    if is_superuser:
        return list(ALL_MODULE_KEYS)
    role_def = resolve_role(role)
    role_mods = role_def.get('modules')
    if role_mods is None:
        role_mods = list(ALL_MODULE_KEYS)
    tenant = list(tenant_modules or [])
    if not tenant:
        # Empty tenant list historically means "all enabled" in this ERP
        return list(role_mods)
    if role_def['key'] in ('admin', 'superadmin'):
        return [m for m in tenant if m in ALL_MODULE_KEYS] or list(tenant)
    allowed = set(role_mods)
    return [m for m in tenant if m in allowed]


def role_allows_section(role: str | None, section_title: str, *, is_superuser: bool = False) -> bool:
    if is_superuser:
        return True
    role_def = resolve_role(role)
    if role_def['key'] in ('admin', 'superadmin'):
        return True
    sections = role_def.get('menu_sections')
    if sections is None:
        return True
    if not sections:
        return section_title == 'Dashboard'
    return section_title in sections or section_title.upper() in {s.upper() for s in sections}


def has_capability(role: str | None, capability: str, *, is_superuser: bool = False) -> bool:
    if is_superuser:
        return True
    role_def = resolve_role(role)
    if role_def['key'] == 'admin':
        return True
    return capability in (role_def.get('capabilities') or [])


def serialize_rbac(user) -> dict[str, Any]:
    role = getattr(user, 'role', None) or 'staff'
    is_super = bool(getattr(user, 'is_superuser', False))
    tenant = getattr(user, 'tenant', None)
    tenant_mods = tenant.get_enabled_modules() if tenant is not None else []
    role_def = resolve_role(role)
    modules = effective_modules(role, tenant_mods, is_superuser=is_super)
    return {
        'role': role_def['key'],
        'role_label': role_def.get('label') or role,
        'role_description': role_def.get('description') or '',
        'home_path': '/saas' if is_super else (role_def.get('home_path') or '/home'),
        'modules': modules,
        'menu_sections': role_def.get('menu_sections'),
        'capabilities': list(role_def.get('capabilities') or []) if not is_super and role_def['key'] != 'admin' else [
            CAP_VIEW_DASHBOARD,
            CAP_MANAGE_USERS,
            CAP_POST_VOUCHERS,
            CAP_MANAGE_COA,
            CAP_VIEW_FINANCE,
            CAP_PURCHASE,
            CAP_APPROVE_PO,
            CAP_POS,
            CAP_CHECKIN,
            CAP_FOLIO,
            CAP_ROOM_STATUS,
            CAP_HK_TASKS,
        ],
        'color': role_def.get('color') or '#6b7280',
    }


def list_operational_roles() -> list[dict[str, Any]]:
    """Roles shown in login demos / utilities (exclude aliases & superadmin)."""
    keys = [
        'admin',
        'operations_manager',
        'frontdesk',
        'housekeeping',
        'restaurant',
        'accountant',
        'purchase_officer',
    ]
    out = []
    for key in keys:
        d = resolve_role(key)
        out.append({
            'key': key,
            'label': d['label'],
            'description': d['description'],
            'home_path': d['home_path'],
            'color': d.get('color'),
        })
    return out

File: frontend/lib/rbac.ts
/** Frontend RBAC — mirrors backend/api/rbac.py (keep in sync). */

import type { ModuleKey } from './modules'

export type HotelRole =
  | 'superadmin'
  | 'admin'
  | 'operations_manager'
  | 'manager'
  | 'frontdesk'
  | 'housekeeping'
  | 'restaurant'
  | 'fnb'
  | 'accountant'
  | 'purchase_officer'
  | 'maintenance'
  | 'staff'

export interface RoleDef {
  key: HotelRole
  label: string
  description: string
  homePath: string
  modules: ModuleKey[] | null
  menuSections: string[] | null
  color: string
  demoEmail?: string
}

const ALL: ModuleKey[] = [
  'frontdesk',
  'housekeeping',
  'fnb',
  'recipes',
  'laundry',
  'spa',
  'hall',
  'banquet',
  'pool',
  'crm',
  'accounts',
  'inventory',
  'assets',
  'broadcast',
  'hr',
  'channel',
  'reports',
  'utilities',
  'landing',
]

export const ROLE_CATALOG: Record<string, RoleDef> = {
  superadmin: {
    key: 'superadmin',
    label: 'Platform Superadmin',
    description: 'SaaS control plane',
    homePath: '/saas',
    modules: ALL,
    menuSections: null,
    color: '#6366f1',
  },
  admin: {
    key: 'admin',
    label: 'Property Administrator',
    description: 'Full property access',
    homePath: '/home',
    modules: ALL,
    menuSections: null,
    color: '#0f766e',
    demoEmail: 'admin@admin.com',
  },
  operations_manager: {
    key: 'operations_manager',
    label: 'Operations Manager',
    description: 'Rooms, F&B, inventory, assets, CRM, reports',
    homePath: '/home/operations',
    modules: [
      'frontdesk',
      'housekeeping',
      'fnb',
      'recipes',
      'laundry',
      'spa',
      'hall',
      'banquet',
      'pool',
      'crm',
      'inventory',
      'assets',
      'broadcast',
      'reports',
      'channel',
    ],
    menuSections: [
      'Dashboard',
      'FRONTDESK',
      'HOUSEKEEPING',
      'BANQUET',
      'F&B AND REVENUE CENTER',
      'INVENTORY',
      'SALES & MARKETING',
      'ASSET & MAINTANANCE',
      'ASSET & MAINTENANCE',
      'BROADCAST MESSAGE',
      'CHANNEL MANAGER',
      'REPORT CENTER',
    ],
    color: '#1d4ed8',
    demoEmail: 'ops@turag.com',
  },
  manager: {
    key: 'manager',
    label: 'Manager',
    description: 'Operations oversight',
    homePath: '/home/operations',
    modules: null,
    menuSections: null,
    color: '#1d4ed8',
  },
  frontdesk: {
    key: 'frontdesk',
    label: 'Front Desk',
    description: 'Reservations, folio, room rack',
    homePath: '/home/frontdesk',
    modules: ['frontdesk', 'housekeeping', 'crm', 'broadcast', 'reports', 'channel'],
    menuSections: [
      'Dashboard',
      'FRONTDESK',
      'HOUSEKEEPING',
      'SALES & MARKETING',
      'BROADCAST MESSAGE',
      'CHANNEL MANAGER',
      'REPORT CENTER',
    ],
    color: '#2563eb',
    demoEmail: 'frontdesk@turag.com',
  },
  housekeeping: {
    key: 'housekeeping',
    label: 'Housekeeping',
    description: 'Room status, tasks, lost & found',
    homePath: '/home/housekeeping',
    modules: ['housekeeping', 'laundry', 'assets', 'reports'],
    menuSections: [
      'Dashboard',
      'HOUSEKEEPING',
      'LAUNDRY',
      'ASSET & MAINTANANCE',
      'ASSET & MAINTENANCE',
      'REPORT CENTER',
    ],
    color: '#059669',
    demoEmail: 'hk@turag.com',
  },
  restaurant: {
    key: 'restaurant',
    label: 'Restaurant / F&B',
    description: 'POS, menu, kitchen stock',
    homePath: '/home/restaurant',
    modules: ['fnb', 'recipes', 'inventory', 'reports'],
    menuSections: [
      'Dashboard',
      'F&B AND REVENUE CENTER',
      'FOOD & BEVERAGE',
      'RECIPE MANAGEMENT',
      'INVENTORY',
      'REPORT CENTER',
    ],
    color: '#c2410c',
    demoEmail: 'restaurant@turag.com',
  },
  fnb: {
    key: 'fnb',
    label: 'F&B',
    description: 'Restaurant POS',
    homePath: '/home/restaurant',
    modules: null,
    menuSections: null,
    color: '#c2410c',
  },
  accountant: {
    key: 'accountant',
    label: 'Accountant',
    description: 'Ledgers, vouchers, financial reports',
    homePath: '/home/accountant',
    modules: ['accounts', 'reports'],
    menuSections: ['Dashboard', 'ACCOUNTS', 'REPORT CENTER'],
    color: '#7c3aed',
    demoEmail: 'accountant@turag.com',
  },
  purchase_officer: {
    key: 'purchase_officer',
    label: 'Purchase Officer',
    description: 'POs, suppliers, stock — no GL posting',
    homePath: '/home/purchase',
    modules: ['inventory', 'reports'],
    menuSections: ['Dashboard', 'INVENTORY', 'REPORT CENTER'],
    color: '#b45309',
    demoEmail: 'purchase@turag.com',
  },
  maintenance: {
    key: 'maintenance',
    label: 'Maintenance',
    description: 'Assets & work orders',
    homePath: '/home/housekeeping',
    modules: ['assets', 'housekeeping', 'reports'],
    menuSections: [
      'Dashboard',
      'HOUSEKEEPING',
      'ASSET & MAINTANANCE',
      'ASSET & MAINTENANCE',
      'REPORT CENTER',
    ],
    color: '#64748b',
  },
  staff: {
    key: 'staff',
    label: 'Staff',
    description: 'Minimal access',
    homePath: '/home',
    modules: [],
    menuSections: ['Dashboard'],
    color: '#6b7280',
  },
}

const ALIAS: Record<string, string> = {
  manager: 'operations_manager',
  fnb: 'restaurant',
}

export function resolveRole(role?: string | null): RoleDef {
  const key = (role || 'staff').toLowerCase()
  const resolved = ALIAS[key] || key
  const base = ROLE_CATALOG[resolved] || ROLE_CATALOG.staff
  const self = ROLE_CATALOG[key]
  if (self && ALIAS[key]) {
    return { ...base, key: self.key, label: self.label }
  }
  return { ...base, key: (base.key || resolved) as HotelRole }
}

export function homePathForUser(user?: {
  is_superuser?: boolean
  role?: string
  home_path?: string
  rbac?: { home_path?: string }
} | null) {
  if (!user) return '/login'
  if (user.is_superuser) return '/saas'
  return user.home_path || user.rbac?.home_path || resolveRole(user.role).homePath
}

export function roleAllowsMenuSection(
  role: string | undefined | null,
  sectionTitle: string,
  isSuperuser?: boolean
) {
  if (isSuperuser) return true
  const def = resolveRole(role)
  if (def.key === 'admin' || def.key === 'superadmin') return true
  const sections = def.menuSections
  if (sections === null) return true
  if (!sections.length) return sectionTitle === 'Dashboard'
  return sections.some((s) => s.toLowerCase() === sectionTitle.toLowerCase())
}

export function DEMO_ROLE_LOGINS() {
  return [
    ROLE_CATALOG.admin,
    ROLE_CATALOG.operations_manager,
    ROLE_CATALOG.frontdesk,
    ROLE_CATALOG.housekeeping,
    ROLE_CATALOG.restaurant,
    ROLE_CATALOG.accountant,
    ROLE_CATALOG.purchase_officer,
  ].filter((r) => r.demoEmail)
}

/** Map app URL prefixes to SaaS module keys (mirrors API RBAC). */
export function moduleForAppPath(pathname: string): ModuleKey | 'utilities' | null {
  const path = pathname || ''
  if (path.startsWith('/login') || path.startsWith('/saas') || path.startsWith('/site') || path.startsWith('/access-denied') || path.startsWith('/apps')) {
    return null
  }
  if (path.startsWith('/home')) return null
  if (path.startsWith('/accounts')) return 'accounts'
  if (path.startsWith('/inventory')) return 'inventory'
  if (path.startsWith('/fnb')) return 'fnb'
  if (path.startsWith('/housekeeping')) return 'housekeeping'
  if (path.startsWith('/frontdesk') || path.startsWith('/reservations')) return 'frontdesk'
  if (path.startsWith('/banquet')) return 'banquet'
  if (path.startsWith('/hr')) return 'hr'
  if (path.startsWith('/crm') || path.startsWith('/sales')) return 'crm'
  if (path.startsWith('/assets')) return 'assets'
  if (path.startsWith('/broadcast')) return 'broadcast'
  if (path.startsWith('/utilities') || path.startsWith('/website')) return 'utilities'
  if (path.startsWith('/reports')) return 'reports'
  if (path.startsWith('/channel')) return 'channel'
  return null
}

export function userHasModuleAccess(
  user: { is_superuser?: boolean; role?: string; enabled_modules?: string[] } | null | undefined,
  moduleKey: string
) {
  if (!user) return false
  if (user.is_superuser) return true
  const role = (user.role || '').toLowerCase()
  if (role === 'admin' || role === 'superadmin') return true
  const mods = user.enabled_modules
  if (mods === undefined || mods === null) return true
  return mods.includes(moduleKey)
}
